motd/src/lib.rs

```diff
@@ -1,24 +1,234 @@
+        .map_err(|e| merr!("serializing arg_resolver: {}", e))?;
+    let out = Command::new(env::current_exe().map_err(|e| merr!("no exe path: {}", e))?)
+        .env(SO_RESOLVER_ARG_ENV_VAR, so_resolver_json)
+        .env(ARG_RESOLVER_ARG_ENV_VAR, arg_resolver_json)
+        .env("LD_PRELOAD", overlay_so.path())
+        .output()
+        .map_err(|e| merr!("error re-execing self: {}", e))?;
+
+    if !out.status.success() {
+        return Err(merr!("failed to re-exec, bad status = {}", out.status));
+    }
+
+    if out.stderr.len() > 0 {
+        println!("{}", String::from_utf8_lossy(out.stdout.as_slice()));
+
+        let stderr = String::from_utf8_lossy(out.stderr.as_slice());
+        return Err(merr!("in re-execed process: {}", stderr));
+    }
+
+    if out.stdout.len() > 0 {
+        return Ok(String::from_utf8_lossy(out.stdout.as_slice()).into());
+    }
+
+    Err(merr!("no motd output"))
+}
+
+/// You MUST call this routine in the main function of the binary that uses
+/// the motd crate. In order to work around an issue where `pam_motd.so` thinks
+/// it is running as root, we use LD_PRELOAD to stub out some privilege
+/// juggling methods. To do this, the value() routine re-execs the current binary,
+/// and it is in this re-execd process that we actually load and call into
+/// `pam_motd.so`.
+pub fn handle_reexec() {
+    if let (Ok(so_resolver_json), Ok(arg_resolver_json)) = (
+        env::var(SO_RESOLVER_ARG_ENV_VAR),
+        env::var(ARG_RESOLVER_ARG_ENV_VAR),
+    ) {
+        match reexec_call_so(so_resolver_json.as_str(), arg_resolver_json.as_str()) {
+            Ok(motd_msg) => print!("{}", motd_msg),
+            Err(e) => eprintln!("{}", e),
+        }
+        std::process::exit(0);
+    }
+}
+
+/// Actually load and call the `pam_motd.so`
+fn reexec_call_so(so_resolver_json: &str, arg_resolver_json: &str) -> Result<String, Error> {
+    let so_resolver: PamMotdResolutionStrategy = serde_json::from_str(so_resolver_json)
+        .map_err(|e| merr!("parsing so_resolver arg: {}", e))?;
+    let arg_resolver: ArgResolutionStrategy = serde_json::from_str(arg_resolver_json)
+        .map_err(|e| merr!("parsing arg_resolver arg: {}", e))?;
+
+    let mut conv_data = ConvData::new();
+    let pam_conv = pam_sys::types::PamConversation {
+        conv: Some(conv_handler),
+        // Safety: It is always safe to cast to void in the immediate moment,
+        //         and we will be done with the conversation by the time this
+        //         routine returns and removes the underlying allocation from
+        //         the stack.
+        data_ptr: unsafe { mem::transmute::<_, *mut libc::c_void>(&mut conv_data) },
+    };
+
+    // Safety: pretty much pure ffi, the errono access follows the instructions documented
+    //         in man getpwuid.
+    let passwd = unsafe {
+        *libc::__errno_location() = 0;
+        let passwd = libc::getpwuid(libc::getuid());
+        let errno = *libc::__errno_location();
```

Comment:
Oh nice, I didn't know about that API. I kept the existing __errno_location accesses since I still need to zero it out ahead of time and I can't tell how to check if the os error returned by last_os_error means "no error". It seems like the kind of value that is only supposed to exit within error conditions.